### PR TITLE
Add CUDA ProjectionProfile Implementation

### PR DESCRIPTION
**Pull request by josephnicholas**
### Changes
- Add CUDA implementation for ProjectionProfile

https://github.com/ihhub/penguinV/issues/371

### CODE REVIEW

**Comment by josephnicholas** [comment]
@ihhub any thoughts on this PR? 💯

File: src/cuda/image_function_cuda.cu [hunk 1]
```diff
@@ -43,6 +44,16 @@ namespace
     const FunctionRegistrator functionRegistrator;
 
     // The list of CUDA device functions on device side
+    __global__ void normalizeCuda( const uint8_t * image, uint32_t rowSize, bool horizontal, uint32_t width, uint32_t height, uint32_t * projection )
```
Comment (ihhub) [inline src/cuda/image_function_cuda.cu]: Please put this function at the place to have alphabetic order among all functions

Comment (ihhub) [reply src/cuda/image_function_cuda.cu]: Why is it called `normalize` function?

File: src/cuda/image_function_cuda.cu [hunk 2]
```diff
@@ -1170,4 +1181,40 @@ namespace Image_Function_Cuda
...
+        std::fill( projection.begin(), projection.end(), 0u );
+
+        const uint32_t rowSize = image.rowSize();
+        width *= colorCount;
+
+        const uint8_t * imageX = image.data() + y * rowSize + x * colorCount;
+
+        multiCuda::Array< uint32_t > projectionCuda( projection );
+
+        launchKernel2D( normalizeCuda, width, height,
```
Comment (ihhub) [inline src/cuda/image_function_cuda.cu]: This line of code should call `normalizeCuda` function.

File: src/cuda/image_function_cuda.cu [hunk 3]
```diff
@@ -1170,4 +1181,40 @@ namespace Image_Function_Cuda
...
+            imageX, rowSize, horizontal, width, height, projectionCuda.data());
+    }
```
Comment (ihhub) [inline src/cuda/image_function_cuda.cu]: We should copy `projectionCuda` into `projection` array upon return from this function.

File: src/cuda/image_function_cuda.cu [hunk 4]
```diff
@@ -43,6 +44,16 @@ namespace
     const FunctionRegistrator functionRegistrator;
 
     // The list of CUDA device functions on device side
+    __global__ void normalizeCuda( const uint8_t * image, uint32_t rowSize, bool horizontal, uint32_t width, uint32_t height, uint32_t * projection )
+    {
+        const uint32_t x = blockDim.x * blockIdx.x + threadIdx.x;
+        const uint32_t y = blockDim.y * blockIdx.y + threadIdx.y;
+
+        if ( x < width && y < height ) {
```
Comment (ihhub) [inline src/cuda/image_function_cuda.cu]: This function is not completed.

**Comment by josephnicholas** [comment]
Will do the changes sir

File: src/cuda/image_function_cuda.cu (hunk 4 above)
Comment (josephnicholas) [reply src/cuda/image_function_cuda.cu]: Should I refer to the OpenCL implementation on `projectionHorizontalOpenCL` and `projectionVerticalOpenCL` based on the `horizontal` value?

File: src/cuda/image_function_cuda.cu (hunk 1 above)
Comment (josephnicholas) [reply src/cuda/image_function_cuda.cu]: changing the function name.

File: src/cuda/image_function_cuda.cu (hunk 2 above)
Comment (josephnicholas) [reply src/cuda/image_function_cuda.cu]: This line is already calling `normalizeCuda` or do you mean it should only call the function.

**Comment by josephnicholas** [comment]
@ihhub any suggestions on my questions?

File: src/cuda/image_function_cuda.cu (hunk 4 above)
Comment (ihhub) [reply src/cuda/image_function_cuda.cu]: Yes.

File: src/cuda/image_function_cuda.cu (hunk 1 above)
Comment (ihhub) [reply src/cuda/image_function_cuda.cu]: It should be called something like `projectionProfileCuda`.

**Comment by ihhub** [comment]
Hi @josephnicholas I put some comments in pull request. Could you please take a look into them?

**Comment by ihhub** [comment]
Hi @josephnicholas why did you close your pull request?

**Comment by josephnicholas** [comment]
Sorry I was gonna create another clean PR, conflicts were hard to correct.

**Comment by josephnicholas** [comment]
@ihhub closing this PR since I can't change source branch. I will continue development on this branch.

**Pull request closed without merging**